src/frontend/src/optimizer/rule/table_function_to_internal_backfill_progress.rs

```diff
@@ -0,0 +1,184 @@
+// Copyright 2025 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::rc::Rc;
+use std::sync::Arc;
+
+use anyhow::bail;
+use itertools::Itertools;
+use risingwave_common::catalog::{Field, Schema, is_backfill_table};
+use risingwave_common::types::{DataType, ScalarImpl};
+use risingwave_expr::aggregate::AggType;
+pub use risingwave_pb::expr::agg_call::PbKind as PbAggKind;
+
+use super::{ApplyResult, BoxedRule, FallibleRule};
+use crate::TableCatalog;
+use crate::catalog::catalog_service::CatalogReadGuard;
+use crate::expr::{AggCall, ExprImpl, InputRef, Literal, OrderBy, TableFunctionType};
+use crate::optimizer::plan_node::generic::GenericPlanRef;
+use crate::optimizer::plan_node::{
+    LogicalAgg, LogicalProject, LogicalScan, LogicalTableFunction, LogicalUnion, LogicalValues,
+};
+use crate::optimizer::{OptimizerContext, PlanRef};
+use crate::utils::{Condition, GroupBy};
+
+/// Transform the `internal_backfill_progress()` table function
+/// into a plan graph which will scan the state tables of backfill nodes.
+/// It will return the progress of the backfills, partitioned by the backfill node's fragment id.
+pub struct TableFunctionToInternalBackfillProgressRule {}
+impl FallibleRule for TableFunctionToInternalBackfillProgressRule {
+    fn apply(&self, plan: PlanRef) -> ApplyResult {
+        let logical_table_function: &LogicalTableFunction = plan.as_logical_table_function()?;
+        if logical_table_function.table_function.function_type
+            != TableFunctionType::InternalBackfillProgress
+        {
+            return ApplyResult::NotApplicable;
+        }
+
+        let reader = plan.ctx().session_ctx().env().catalog_reader().read_guard();
+        // TODO(kwannoel): Make sure it reads from source tables as well.
+        let backfilling_tables = get_backfilling_tables(reader);
+        let plan = Self::build_plan(plan.ctx(), backfilling_tables)?;
+        ApplyResult::Ok(plan)
+    }
+}
+
+impl TableFunctionToInternalBackfillProgressRule {
+    fn build_plan(
+        ctx: Rc<OptimizerContext>,
+        backfilling_tables: Vec<Arc<TableCatalog>>,
+    ) -> anyhow::Result<PlanRef> {
+        if backfilling_tables.is_empty() {
+            let fields = vec![
+                Field::new("job_id", DataType::Int32),
```

Comment:
Does field name matter here? It seems that the two branches (the returned `LogicalValues` and `LogicalUnion`) have different field names.

Reply:
No It's just for readability. There's a testcase here which [tests both paths](https://github.com/risingwavelabs/risingwave/pull/22124/files#diff-5ecb1675d8a5f20dc204a17255307abb3b4e95ee71846bd8d0db2d8f1c51578eR11).

1. Backfill Tables are empty
2. Backfill Tables are non-empty

Reply:
The test is further [enhanced here](https://github.com/risingwavelabs/risingwave/pull/22166/files#diff-0d1aad88fbcb6272b3c47d81410826a8fd02c2b7545e197f2cd36273dbb148cc) to do regex checks on progress.